lib/streamlit/runtime/session_manager.py

```diff
@@ -0,0 +1,344 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import Dict, List, Optional, TYPE_CHECKING
+
+from typing_extensions import Protocol
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+from .app_session import AppSession
+
+if TYPE_CHECKING:
+    from .runtime import Runtime
+
+
+class SessionClientDisconnectedError(Exception):
+    """Raised by operations on a disconnected SessionClient."""
+
+
+class SessionClient(Protocol):
+    """Interface for sending data to a session's client."""
+
+    @abstractmethod
+    def write_forward_msg(self, msg: ForwardMsg) -> None:
+        """Deliver a ForwardMsg to the client.
+
+        If the SessionClient has been disconnected, it should raise a
+        SessionClientDisconnectedError.
+        """
+        raise NotImplementedError
+
+
+# TODO(vdonato): We may want to consider either coming up with a better name for this or
+# consolidating it with the SessionData class since the two names are similar enough
+# to be confusing right now.
+@dataclass
+class SessionInfo:
+    """Type containing data related to an AppSession.
+
+    For each AppSession, the Runtime tracks that session's
+    script_run_count. This is used to track the age of messages in
+    the ForwardMsgCache.
+    """
+
+    client: Optional[SessionClient]
+    session: AppSession
+    script_run_count: int = 0
+
+
+class SessionStorageError(Exception):
+    """Exception class for errors raised by SessionStorage.
+
+    The original error that causes a SessionStorageError to be (re)raised will generally
+    be an I/O error specific to the concrete SessionStorage implementation.
+    """
+
+
+class SessionStorage(Protocol):
+    @abstractmethod
+    def get(self, session_id: str) -> Optional[SessionInfo]:
+        """Return the SessionInfo corresponding to session_id, or None if one does not
+        exist.
+
+        Parameters
+        ----------
+        session_id
+            The unique ID of the session being fetched.
+
+        Returns
+        -------
+        SessionInfo or None
+
+        Raises
+        ------
+        SessionStorageError
+            Raised if an error occurs while attempting to fetch the session.
+        """
```

Comment:
At first glance, this is a bit unclear. Since "session id does not exist" is not an error, what sorts of issues could occur when attempting to fetch a session? Can we provide an example in the docstring, like:

```
"""
SessionStorageError
    Raised if an error occurs while attempting to fetch the session (e.g. the storage backend is disconnected).
"""
```

Reply:
👍